### ISSUE TITLE
About train domain encoder with 1024 resolution?

### ISSUE THREAD

**Issue by GreenLimeSia** [bryandlee/stylegan2-encoder-pytorch#2]
I try to train the encoder with 1024 FFHQ data, and I do it on Tesla V100 with 800000 iterations, but the result looks not good. I have some questions about how to train the encoder with 1024 resolution?  1、Why use the weights in VGG19 perceptual loss?
2、Which is a good candidate for L1 or L2 loss to train the encoder in 1024 resolution?  3、Should we increase the number of iterations to improve the results? 

Note: I upsample the image from 1024 to 256 in order to satisfy the perceptual loss, and I change the output ((14,512) for 256 resolution) to (18, 512) to satisfy the input of the generator.

would you mind gei me some advice to solve this issue?

> class Encoder(nn.Module):
    def __init__(self, size, w_dim=512):
        super().__init__()

        channels = {
            4: 512,
            8: 512,
            16: 512,
            32: 512,
            64: 256,
            128: 128,
            256: 64,
            512: 32,
            1024: 16
        }

        self.w_dim = w_dim
        log_size = int(math.log(size, 2))

        self.n_latents = 18  # log_size * 2 - 2

> recon_img, _ = generator([latents],truncation=truncation,truncation_latent=trunc,randomize_noise=False)

        recon_img_resize = recon_img.clone()

        recon_img_resize = upsample2d(recon_img_resize)

        recon_vgg_loss = vgg_loss(recon_img_resize, real_img_resize)

Results image：
![795900](https://user-images.githubusercontent.com/28706611/94634627-18c0b600-0303-11eb-9bd4-9163d1b1c7b2.png)

**Comment by bryandlee** [comment]
Hi, thanks for sharing the result in 1024!

1. The weights are used to control the relative contribution of each feature layer to the loss. I used the weight from the perceptual loss in [SPADE](https://github.com/NVlabs/SPADE/blob/1a687baeada266a3c92be41295f1ea3d5efd4f93/models/networks/loss.py#L102) paper, whereas the original domain encoder implementation only uses conv4_3 (layer23) of VGG16. You can choose either one, but the more important thing is to **use the original image size for the perceptual loss**. It only uses the convolutional part of VGG network, so you don't have to resize the image. Downsampling the image will hurt the performance in terms of fine details.
2. The original implementation uses L2 loss, but L1 also works well and I can't really tell the difference between them.
3. More iterations surely help.

The original work of the authors was done in 256-scale images, and I'm not pretty sure the architecture of the encoder is well suited for larger images such as 1024. Perhaps the hierarchical architecture proposed in https://github.com/genforce/ghfeat could help.

**Comment by GreenLimeSia** [comment]
I am so sorry for the late reply. I have a doubt about this question **use the original image size for the perceptual loss**.  In the paper "Image2StyleGAN: How to Embed Images Into the StyleGAN Latent Space?", it refers to **limitation of the ImageNet-based Perceptual loss.**  what do you think?
![1601909112(1)](https://user-images.githubusercontent.com/28706611/95095272-6dc65700-075d-11eb-86af-ca9a54542f18.jpg)

**Comment by bryandlee** [comment]
There could be a tradeoff between the fine details and scale-dependent features learned in vgg. 
I've tested the vgg loss module by directly optimizing the pixels values at 1024 with/without resizing, and the results are somewhat off in both cases.
(The results might be different when combined with mse loss and optimized in the generator latent space.)

(1) target 1024 image
![image](https://user-images.githubusercontent.com/26464535/95143847-29b17180-07b2-11eb-972e-3160f88e1fe2.png) 
(2) 1024 image optimized with vgg loss at 1024
![image](https://user-images.githubusercontent.com/26464535/95143864-35049d00-07b2-11eb-98b5-1f6440ac7d18.png) 
(3) 1024 image optimized with vgg loss at 256
![image](https://user-images.githubusercontent.com/26464535/95143873-3cc44180-07b2-11eb-984a-c718c90b3647.png) 
(4) 256 downsampled image of (3)
![image](https://user-images.githubusercontent.com/26464535/95143870-3930ba80-07b2-11eb-913b-6deea188593c.png)

Also, the uniform vgg weights used in Image2StyleGAN seem to work better than the ones in pix2pix-related codes.

**Comment by GreenLimeSia** [comment]
Thanks for your hard work! I will follow you without resizing the image. I will train the encoder again to verify the power of the encoder architecture.

**Comment by bryandlee** [comment]
Fyi, the original authors said they are going to upload a full-functioning pytorch code as well!

https://github.com/genforce/idinvert_pytorch/issues/3#issuecomment-664216404
